### PR TITLE
drivers: eth_mcux: adding i.mx-rt support

### LINKED ISSUE #8900: Enable mcux ethernet driver on iMX RT1050

### PR DESCRIPTION
**Pull request by agansari**
Enables Networking hardware on i.MX-RT 1050-EVKB board.
Fixes #8900
 - enables board specific pinout for ethernet connection
 - fixed initialization sequence in eth_mcux
 - pinmux enablement pushed later so sysclock is initalized before

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

### CODE REVIEW

File: drivers/ethernet/eth_mcux.c [hunk 1]
```diff
@@ -932,6 +949,38 @@ static void eth_mcux_ptp_isr(void *p)
...
+#ifdef CONFIG_SOC_SERIES_IMX_RT
+static void eth_mcux_dispacher_isr(void *p)
+{
+	struct device *dev = p;
+	struct eth_context *context = dev->driver_data;
+	uint32_t EIR = ENET_GetInterruptStatus(ENET);
+	int irq_lock_key = irq_lock();
+
+
+	if (EIR & (kENET_RxBufferInterrupt | kENET_RxFrameInterrupt))
```
Comment (jukkar) [inline drivers/ethernet/eth_mcux.c]: This does not follow Zephyr coding style. Please place { in the same line as the if() statement like this
```
if (xxx) {

} else if () {

} else {

}
```

File: drivers/ethernet/eth_mcux.c [hunk 2]
```diff
@@ -932,6 +949,38 @@ static void eth_mcux_ptp_isr(void *p)
...
+		ENET_ClearInterruptStatus(ENET, kENET_MiiInterrupt);
+	}
+	else if (EIR)
+	{
+		ENET_ClearInterruptStatus(ENET, 0xFFFFFFFF);
+	}
+
+	irq_unlock(irq_lock_key);
+}
+#endif //CONFIG_SOC_SERIES_IMX_RT
```
Comment (jukkar) [inline drivers/ethernet/eth_mcux.c]: No // comments

**Review by jukkar — CHANGES_REQUESTED** [review]

**Review by jukkar — APPROVED** [review]
LGTM, but please check the issues sanitychecker is reporting at https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/25592/5/tests

File: boards/arm/mimxrt1050_evk/pinmux.c [hunk 3]
```diff
@@ -65,27 +74,63 @@ static int mimxrt1050_evk_init(struct device *dev)
...
+	GPIO_WritePinOutput(GPIO1, 9, 0);
+	k_busy_wait(100 * USEC_PER_MSEC);
+	GPIO_WritePinOutput(GPIO1, 9, 1);
 #endif
 
 	return 0;
 }
 
-SYS_INIT(mimxrt1050_evk_init, PRE_KERNEL_1, 0);
+SYS_INIT(mimxrt1050_evk_init, PRE_KERNEL_2, 0);
```
Comment (galak) [inline boards/arm/mimxrt1050_evk/pinmux.c]: Curious, why the change to PRE_KERNEL_2

**Review by galak — CHANGES_REQUESTED** [review]
Can you fixup the white space issues that checkpatch is reporting
Also, can you add more details about what changes you are making to the ethernet driver itself to support i.MX-RT (i'd pull these changes out into their own commit).

File: boards/arm/mimxrt1050_evk/pinmux.c (hunk 3 above)
Comment (agansari) [reply boards/arm/mimxrt1050_evk/pinmux.c]: Pushed pin initialization later so i can use k_busy_wait to reset ethernet PHY chip.

**Comment by pfalcon** [comment]
From commit message:

> Enables Newtworking hardware on i.MX-RT 1050-EVKB board.

Please fix typo in "Newtworking".

And the commit message is very bare. Even description of the PR here more detailed than the commit message, even though the commit message is what goes into posterity (into git log, git bisect, etc.). At the very least, it could mention that existing eth_mcux driver is reused for this hardware (which is not obvious and pleasantly (I hope) surprising fact for me).

(Note that "more detailed commit message" is my usual suggestion to pretty big share of patches I review.)

File: boards/arm/mimxrt1050_evk/Kconfig.defconfig [hunk 4]
```diff
@@ -54,4 +54,14 @@ config TEXT_SECTION_OFFSET
 
 endif
 
+if NETWORKING
+
+config NET_L2_ETHERNET
+	def_bool y if !MODEM_WNCM14A2A
```
Comment (pfalcon) [inline boards/arm/mimxrt1050_evk/Kconfig.defconfig]: Is this truly a good idea here?

File: boards/arm/mimxrt1050_evk/pinmux.c [hunk 5]
```diff
@@ -65,27 +74,63 @@ static int mimxrt1050_evk_init(struct device *dev)
 	IOMUXC_SetPinMux(IOMUXC_GPIO_AD_B0_03_LPSPI3_PCS0, 0);
 
 	IOMUXC_SetPinConfig(IOMUXC_GPIO_AD_B0_00_LPSPI3_SCK,
-			    IOMUXC_SW_PAD_CTL_PAD_PKE_MASK |
-			    IOMUXC_SW_PAD_CTL_PAD_SPEED(2) |
-			    IOMUXC_SW_PAD_CTL_PAD_DSE(6));
+				IOMUXC_SW_PAD_CTL_PAD_PKE_MASK |
```
Comment (pfalcon) [inline boards/arm/mimxrt1050_evk/pinmux.c]: So, in such cases, Zephyr codestyle requires to align args "in column", using tabs for as long as possible, then spaces. In other words, what was before seems to be right. Also, IMHO, codestyle changes should be separated from functional changes. YMMV

File: boards/arm/mimxrt1050_evk/pinmux.c [hunk 6]
```diff
@@ -65,27 +74,63 @@ static int mimxrt1050_evk_init(struct device *dev)
...
+				IOMUXC_SW_PAD_CTL_PAD_SPEED(2) |
+				IOMUXC_SW_PAD_CTL_PAD_DSE(6));
+#endif
+
+#ifdef CONFIG_ETH_MCUX_0
+	IOMUXC_SetPinMux(IOMUXC_GPIO_B1_04_ENET_RX_DATA00, 0U);
```
Comment (pfalcon) [inline boards/arm/mimxrt1050_evk/pinmux.c]: Why do we have uppercase "U" here and "u" below?

Also, so far Zephyr doesn't use that "suffixed U" style. (So far, yes).

**Review by pfalcon — DISMISSED** [review]
Some suggestions from my side too. At least splitting off ext/ change to a separate commit should be done. (Unless overridden by @galak, @MaureenHelm or someone else in the know.)

File: drivers/ethernet/eth_mcux.c [hunk 7]
```diff
@@ -932,6 +951,30 @@ static void eth_mcux_ptp_isr(void *p)
 }
 #endif
 
+#ifdef CONFIG_SOC_SERIES_IMX_RT
```
Comment (MaureenHelm) [inline drivers/ethernet/eth_mcux.c]: I'd like to avoid direct dependencies on the SoC like this. If you add `interrupt-names = "COMMON"` to the dts, you can make this condition `#ifdef ETH_IRQ_COMMON` instead.

File: drivers/ethernet/eth_mcux.c [hunk 8]
```diff
@@ -932,6 +951,30 @@ static void eth_mcux_ptp_isr(void *p)
...
+		ENET_ClearInterruptStatus(ENET, kENET_MiiInterrupt);
+	} else if (EIR) {
+		ENET_ClearInterruptStatus(ENET, 0xFFFFFFFF);
+	}
+
+	irq_unlock(irq_lock_key);
+}
+#endif /* CONFIG_SOC_SERIES_IMX_RT */
+
+#ifdef CONFIG_SOC_SERIES_KINETIS_K6X
```
Comment (MaureenHelm) [inline drivers/ethernet/eth_mcux.c]: Change this to `#ifdef ETH_IRQ_RX` and wrap the other isr's similarly with `#ifdef ETH_IRQ_TX` and `#ifdef ETH_IRQ_ERR_MISC`

File: drivers/ethernet/eth_mcux.c [hunk 9]
```diff
@@ -993,6 +1038,12 @@ static void eth_0_config_func(void)
 	IRQ_CONNECT(CONFIG_IRQ_ETH_ERR_MISC, CONFIG_ETH_MCUX_0_IRQ_PRI,
 		    eth_mcux_error_isr, DEVICE_GET(eth_mcux_0), 0);
 	irq_enable(CONFIG_IRQ_ETH_ERR_MISC);
+#endif
+#if CONFIG_SOC_SERIES_IMX_RT
+	IRQ_CONNECT(ENET_IRQn, CONFIG_ETH_MCUX_0_IRQ_PRI,
+		    eth_mcux_dispacher_isr, DEVICE_GET(eth_mcux_0), 0);
+	irq_enable(ENET_IRQn);
```
Comment (MaureenHelm) [inline drivers/ethernet/eth_mcux.c]: Use the irq number from dts instead of CMSIS

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Please make the driver changes in a separate commit from the soc and board changes.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/ethernet/eth_mcux.c [hunk 10]
```diff
@@ -33,7 +33,9 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 
 #include "fsl_enet.h"
 #include "fsl_phy.h"
+#ifdef CONFIG_SOC_SERIES_KINETIS_K6X
 #include "fsl_port.h"
```
[the remark this answers is not in the document]
Comment (agansari) [reply drivers/ethernet/eth_mcux.c]: It was needed in case of unique mac, SIM definition was not included.
No longer required.

File: dts/arm/nxp/nxp_rt.dtsi [hunk 11]
```diff
@@ -30,15 +30,15 @@
 			#size-cells = <1>;
 
 			itcm0: itcm@0 {
-				reg = <0x00000000 0x20000>;
+				reg = <0x00000000 0x80000>;
```
[the remark this answers is not in the document]
Comment (agansari) [reply dts/arm/nxp/nxp_rt.dtsi]: My bad, i just looked at memory map and not FlexRam.
Thank you for yaml suggestion.

**Comment by agansari** [comment]
@pfalcon and @galak can you please help me test this pull request on both i.MX RT and K64F for backward compatibility?

File: drivers/ethernet/eth_mcux.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (MaureenHelm) [reply drivers/ethernet/eth_mcux.c]: Let's remove it then

File: drivers/ethernet/eth_mcux.c [hunk 12]
```diff
@@ -33,9 +33,6 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 
 #include "fsl_enet.h"
 #include "fsl_phy.h"
-#ifdef CONFIG_SOC_SERIES_KINETIS_K6X
-#include "fsl_port.h"
```
Comment (MaureenHelm) [inline drivers/ethernet/eth_mcux.c]: Move this change to the previous commit

File: boards/arm/mimxrt1050_evk/mimxrt1050_evk.yaml [hunk 13]
```diff
@@ -15,3 +15,4 @@ ram: 128
 flash: 128
 supported:
   - spi
+  - netif:eth
```
Comment (MaureenHelm) [inline boards/arm/mimxrt1050_evk/mimxrt1050_evk.yaml]: Could you also update the board doc to reflect that ethernet is now supported?

**Review by MaureenHelm — COMMENTED** [review]
Looks pretty good. I still need to test on hardware

**Review by dbkinder — APPROVED** [review]
+1 for doc changes

**Review by MaureenHelm — APPROVED** [review]

**Comment by galak** [comment]
> Some suggestions from my side too. At least splitting off ext/ change to a separate commit should be done. (Unless overridden by @galak, @MaureenHelm or someone else in the know.)

I don't think it needs its own commit

**Comment by galak** [comment]
@pfalcon still have issues?

**Comment by pfalcon** [comment]
Dismissed my -1, sorry, completely forgot I had it there.

(Submitters, feel free to ping reviewers in such cases.)

File: soc/arm/nxp_imx/rt/dts_fixup.h [hunk 14]
```diff
@@ -41,4 +41,13 @@
 #define DT_UART_MCUX_LPUART_3_CLOCK_NAME	DT_NXP_KINETIS_LPUART_4018C000_CLOCK_CONTROLLER
 #define DT_UART_MCUX_LPUART_3_CLOCK_SUBSYS	DT_NXP_KINETIS_LPUART_4018C000_CLOCK_NAME
 
+#define DT_ETH_MCUX_0_NAME		ETH_LABEL
```
Comment (galak) [inline soc/arm/nxp_imx/rt/dts_fixup.h]: Would prefer to see this be:

#define DT_ETH_MCUX_0_NAME		DT_NXP_KINETIS_ETHERNET_402D8000_LABEL

Comment (agansari) [reply soc/arm/nxp_imx/rt/dts_fixup.h]: @galak I've made the change you requested
@galak @pfalcon Thank you for your reviews!

Comment (pfalcon) [reply soc/arm/nxp_imx/rt/dts_fixup.h]: Sadly, I don't have much to be thanked for - it's on my TODO to test this patch on real hw, but unfortunately I can't to it due to other tasks I got into. I just hope that other folks tested it. (I'll get to it to of course.)

Comment (agansari) [reply soc/arm/nxp_imx/rt/dts_fixup.h]: It's taking me some time to test i.MX RT and backward checking K64F, this i why I didn't ping earlier. Patch looks okay so far.

**Comment by pfalcon** [comment]
@agansari : FYI, I'm very, very close to help with testing this (at least on frdm_k64f). I actually hoped to do this today, but as usually happens, discovered a trail of regressions on approaching that. For you reference, my standard test plans are described in https://github.com/pfalcon/zephyr/wiki/NetworkingTestPlan (again, they don't work with current master, and require patches - new and old).

**Comment by pfalcon** [comment]
Ok, testing: frdm_k64f with samples/net/sockets/dumb_http_server,  master 57a6858278fbe4c180f98c2316d992207336040e + this patch + https://github.com/zephyrproject-rtos/zephyr/pull/7831, as required for reliable frdm_k64f operation (not related to the changes in this PR in any way, master is broken).

`ab -n1000 http://192.0.2.1:8080/` works as expected. Specifically:
~~~
Complete requests:      1000
Failed requests:        0
Total transferred:      2181000 bytes
HTML transferred:       2122000 bytes
Requests per second:    118.71 [#/sec] (mean)
Time per request:       8.424 [ms] (mean)
Time per request:       8.424 [ms] (mean, across all concurrent requests)
Transfer rate:          252.85 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    1   0.1      1       2
Processing:     5    8   0.2      8      11
Waiting:        4    7   0.2      7       7
Total:          6    8   0.2      8      11

Percentage of the requests served within a certain time (ms)
  50%      8
  66%      8
  75%      8
  80%      8
  90%      8
  95%      9
  98%      9
  99%      9
 100%     11 (longest request)
~~~

**Comment by pfalcon** [comment]
Ok, now frdm_k64f + big_http_download:

First iteration went ok:
~~~
***** Booting Zephyr OS zephyr-v1.13.0-2040-g4b22b4c6ce *****
Preparing HTTP GET request for http://archive.ubuntu.com:80/ubuntu/dists/xenial/main/installer-amd64/current/images/hd-media/vmlinuz
addrinfo @0x20008238: ai_family=2, ai_socktype=1, ai_protocol=6, sa_family=2, sin_port=5000
sock = 0
7015896 bytes
Hash: 337c37d7ec00348414224baa6bdb2d43f2a34ef5676bafcdcad916f148b5b317
Total downloaded so far: 6MB
~~~

It seems to be even faster with this patch ;-). (I'm kidding, it's as slow as usual. In all fairness, I don't add d/l speed calculation to not scare off people.)

But on 2nd iter:

~~~
sock = 0
[00:00:01.642,867] <err> eth_mcux.eth_rx: Failed to get fragment buf
1158170 bytes
~~~

Well, that unlikely has anything to do with the driver changes, and should be a transient error.

**Comment by pfalcon** [comment]
Ok, as there're no regressions with frm_k64f, let me approve this. Otherwise, I of course take a chance to bootstrap me with mimxrt1050_evk testing.

**Review by pfalcon — APPROVED** [review]
Tested on frdm_k64f

**Comment by agansari** [comment]
@pfalcon i.MX RT will have some issues, some samples are too big for iTCM memory, but thye will run if they fit.
Thank you for taking time on regressive testing. Can we automate these tests?

**Review by tbursztyka — APPROVED** [review]

**Comment by pfalcon** [comment]
> Can we automate these tests?

Well, that's definitely the aim various parties want to achieve, e.g. I know that NXP has some internal system (@dleach02 told), on Linaro level we want to integrate more HW testing into mostly barebones testing, Zephyr recently added real-board testing into its "sanitycheck" testsuite. But what can I say more if the current situation is that mainline has networking stack setting which semi-break frmd_k64f (https://github.com/zephyrproject-rtos/zephyr/pull/7831), and for half a year (it's actually longer than that) it can't come to resolution?

All in all, before we'll be able to run tests automatically, we need to get used to run tests consistently manually. And https://github.com/pfalcon/zephyr/wiki/NetworkingTestPlan is my modest contribution towards that.

**Comment by pfalcon** [comment]
@MaureenHelm, on https://www.nxp.com/support/developer-resources/run-time-software/i.mx-developer-resources/i.mx-rt1050-evaluation-kit:MIMXRT1050-EVK?tab=In-Depth_Tab (which is click away from "MIMXRT1050-EVK Website" link in https://docs.zephyrproject.org/latest/boards/arm/mimxrt1050_evk/doc/mimxrt1050_evk.html), clicking on "ARM mbed" pane leads to https://os.mbed.com/platforms/FRDM-KW41Z/ , instead of https://os.mbed.com/platforms/MIMXRT1050-EVK/ . Let me know if I should submit this issue in another way ;-).

**Comment by pfalcon** [comment]
@agansari, Maybe you can confirm something to me. So, when I work with a board which has mbed bootloader, I usually follow the least-resistance way, and just copy zephyr.bin files to mass-storage device the said bootloader offers. So far, that doesn't work with mimxrt1050_evk. Symptoms are as described here: https://community.nxp.com/thread/486834 , and I wonder if that reply nails:

>  For I.MXRT1050 EVKB, Drag and drop function is not supported.

Did you ever use the I.MXRT1050 in mbed-like way?

**Comment by agansari** [comment]
@pfalcon I only used j-link debugging, haven't tried mbed bootloader, let me test that.

**Review by galak — APPROVED** [review]

**Comment by pfalcon** [comment]
@agansari: Ok, "mystery" solved. With default-as-shipped mbet bootloader, there's no jlink support. After jlink firmware installed, there's no mbed flash-drive any longer. I was able to get hello_world running all in all. Will continue with eth testing tomorrow. And hope to submit update to the board doc.

**Comment by pfalcon** [comment]
@agansari : So far, bad news - no response from Eth on imxrt, nothing in stats after pings. What I noticed is that both LEDs on Eth jack always light up for me, even without cable. And they never blink. Do you have the same?

And just reinsuring - I didn't see all the jumpers on the board being described in user guide. I don't need change any jumpers, do I?

**Comment by agansari** [comment]
@pfalcon yes, the leds light up from the begining, but they blink on pings for example. There is no need for jumpers or deep switches changes.

**Comment by pfalcon** [comment]
Ok, let's continue in https://github.com/zephyrproject-rtos/zephyr/issues/11586
